libs/ml/examples/distributed_learning/main.cpp

```diff
@@ -0,0 +1,420 @@
+  }
+
+  /**
+   * Adds own gradient to peers queues
+   */
+  void BroadcastGradients()
+  {
+    // Load own gradient
+    TensorVectorType current_gradient = g_.GetGradients();
+
+    // Give gradients to peers
+    for (unsigned int i(0); i < number_of_peers_; ++i)
+    {
+      peers_[i]->AddGradient(current_gradient);
+    }
+  }
+
+  /**
+   * Adds gradient to own gradient queue
+   * @param gradient
+   */
+  void AddGradient(TensorVectorType gradient)
+  {
+    {
+      std::lock_guard<std::mutex> l(queue_mutex_);
+      gradient_queue.push(gradient);
+    }
+  }
+
+  /**
+   * Applies gradient multiplied by -LEARNING_RATE
+   * @param gradients
+   */
+  void ApplyGradient(TensorVectorType gradients)
+  {
+    // Step of SGD optimiser
+    for (SizeType j{0}; j < gradients.size(); j++)
+    {
+      fetch::math::Multiply(gradients.at(j), -LEARNING_RATE, gradients.at(j));
+    }
+
+    // Apply gradients to own model
+    {
+      std::lock_guard<std::mutex> l(model_mutex_);
+      g_.ApplyGradients(gradients);
+    }
+  }
+
+  /**
+   * Rewrites current model with given one
+   * @param new_weights Vector of weights that represent model
+   */
+  void SetWeights(TensorVectorType &new_weights)
+  {
+    std::lock_guard<std::mutex> l(const_cast<std::mutex &>(model_mutex_));
+    g_.set_weights(new_weights);
+  }
+
+private:
+  // Client's own graph and mutex to protect it's weights
+  fetch::ml::Graph<TensorType> g_;
+  std::mutex                   model_mutex_;
+
+  // Client's own dataloader
+  fetch::ml::dataloaders::MNISTLoader<TensorType, TensorType> dataloader_;
+
+  // Connection to other nodes
+  std::vector<std::shared_ptr<TrainingClient>> peers_;
+
+  // Access to coordinator
+  std::shared_ptr<Coordinator> coordinator_ptr_;
+
+  // Gradient queue access and mutex to protect it
+  std::queue<TensorVectorType> gradient_queue;
+  std::mutex                   queue_mutex_;
+
+  // random number generator for shuffling peers
+  fetch::random::LaggedFibonacciGenerator<> gen_;
+
+  // Learning hyperparameters
+  SizeType batch_size_      = BATCH_SIZE;
+  SizeType test_set_ratio_  = TEST_SET_RATIO;
+  SizeType number_of_peers_ = NUMBER_OF_PEERS;
+  SizeType learning_rate_   = LEARNING_RATE
+
+      // Client id (identification name)
+      std::string id_;
+
+  // Timestamp for logging
+  std::string GetTimeStamp()
+  {
+    // TODO: Implement timestamp
+    return "TIMESTAMP";
+  }
+};
+
+int main(int ac, char **av)
+{
+  if (ac < 3)
+  {
+    std::cout << "Usage : " << av[0]
+              << " PATH/TO/train-images-idx3-ubyte PATH/TO/train-labels-idx1-ubyte" << std::endl;
+    return 1;
+  }
+
+  std::shared_ptr<Coordinator> coordinator = std::make_shared<Coordinator>();
+
+  std::cout << "FETCH Distributed MNIST Demo -- Synchronised" << std::endl;
+  std::srand((unsigned int)std::time(nullptr));
+
+  std::vector<std::shared_ptr<TrainingClient>> clients(NUMBER_OF_CLIENTS);
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
```

Comment:
please don't use "unsigned int".
Use either SizeType or a specific uint if you need that (e.g. uint32_t/uint64_t)